### PR TITLE
Key Vault Certificates: added WellKnownIssuerNames enum class

### LINKED ISSUE #8367: Certificates: add WellKnownIssuerNames enum
Should have values "Self" and "Unknown"

### PR DESCRIPTION
**Pull request by iscai-msft**
closes #8367

### CODE REVIEW

**Comment by adxsdk6** [comment]
Can one of the admins verify this patch?

File: sdk/keyvault/azure-keyvault-certificates/azure/keyvault/certificates/client.py [hunk 1]
```diff
@@ -854,7 +855,7 @@ def create_issuer(self, name, provider, **kwargs):
             issuer_attributes = None
         issuer_bundle = self._client.set_certificate_issuer(
             vault_base_url=self.vault_url,
-            issuer_name=name,
+            issuer_name=name.value if not isinstance(name, str) else name,
```
Comment (chlowell) [inline sdk/keyvault/azure-keyvault-certificates/azure/keyvault/certificates/client.py]: The serializer should handle this for you (try it).

File: sdk/keyvault/azure-keyvault-certificates/azure/keyvault/certificates/enums.py [hunk 2]
```diff
@@ -50,3 +50,11 @@ class KeyCurveName(str, Enum):
...
     p_521 = "P-521"  #: The NIST P-521 elliptic curve, AKA SECG curve SECP521R1.
     p_256_k = "P-256K"  #: The SECG SECP256K1 elliptic curve.
+
+
+class WellKnownIssuerNames(str, Enum):
+    """Collection of well-known issuer names"""
+
+    Self = "Self"  #: Use this issuer for a self-signed certificate
+    Unknown = "Unknown"  #: If you use this, you must manually get the cert yourself from the issuer of your choice.
+                         #: Must call :func:`merge_certificate` to complete creation.
```
Comment (chlowell) [inline sdk/keyvault/azure-keyvault-certificates/azure/keyvault/certificates/enums.py]: Does sphinx render the two-line comment properly?

File: sdk/keyvault/azure-keyvault-certificates/azure/keyvault/certificates/models.py [hunk 3]
```diff
@@ -571,8 +569,8 @@ class CertificatePolicy(object):
     :param lifetime_actions: Actions that will be performed by Key Vault over the lifetime
         of a certificate
     :type lifetime_actions: Iterable[~azure.keyvault.certificates.LifetimeAction]
-    :param str issuer_name: Name of the referenced issuer object or reserved names; for example,
-        'Self' or 'Unknown"
+    :param str issuer_name: Name of the referenced issuer object or reserved names. The enum class
+        :class:`~azure.keyvault.certificates.WellKnownIssuerNames` contains popular issuer names.
```
Comment (chlowell) [inline sdk/keyvault/azure-keyvault-certificates/azure/keyvault/certificates/models.py]: Suggestion: ~The enum class~

File: sdk/keyvault/azure-keyvault-certificates/tests/test_examples_certificates_async.py [hunk 4]
```diff
@@ -38,10 +38,12 @@ class TestExamplesKeyVault(AsyncKeyVaultTestCase):
     @AsyncVaultClientPreparer(enable_soft_delete=True)
     @AsyncKeyVaultTestCase.await_prepared_test
     async def test_example_certificate_crud_operations(self, vault_client, **kwargs):
-        certificate_client = vault_client.certificates
+
         # [START create_certificate]
         from azure.keyvault.certificates import CertificatePolicy, SecretContentType
 
+        certificate_client = vault_client.certificates
```
Comment (chlowell) [inline sdk/keyvault/azure-keyvault-certificates/tests/test_examples_certificates_async.py]: :suspect:

File: sdk/keyvault/azure-keyvault-certificates/azure/keyvault/certificates/models.py [hunk 5]
```diff
@@ -793,15 +789,16 @@ def _from_certificate_policy_bundle(cls, certificate_policy_bundle):
             ),
             lifetime_actions=lifetime_actions,
             exportable=key_properties.exportable if key_properties else None,
-            key_type=KeyType(key_properties.key_type) if key_properties and key_properties else None,
+            key_type=KeyType(key_properties.key_type) if key_properties and key_properties.key_type else None,
             key_size=key_properties.key_size if key_properties else None,
             reuse_key=key_properties.reuse_key if key_properties else None,
-            curve=KeyCurveName(key_properties) if key_properties and key_properties.curve else None,
+            curve=KeyCurveName(key_properties.curve) if key_properties and key_properties.curve else None,
```
Comment (chlowell) [inline sdk/keyvault/azure-keyvault-certificates/azure/keyvault/certificates/models.py]: This looks like something that should have failed a test, do we have coverage here?

File: sdk/keyvault/azure-keyvault-certificates/azure/keyvault/certificates/models.py (hunk 3 above)
Comment (iscai-msft) [reply sdk/keyvault/azure-keyvault-certificates/azure/keyvault/certificates/models.py]: Got rid of that, thanks!

File: sdk/keyvault/azure-keyvault-certificates/azure/keyvault/certificates/models.py (hunk 5 above)
Comment (iscai-msft) [reply sdk/keyvault/azure-keyvault-certificates/azure/keyvault/certificates/models.py]: I don't think we actually have a certificate policy that uses a curve, i'll add that to one of them

File: sdk/keyvault/azure-keyvault-certificates/tests/test_examples_certificates_async.py (hunk 4 above)
Comment (iscai-msft) [reply sdk/keyvault/azure-keyvault-certificates/tests/test_examples_certificates_async.py]: moved these out of the coe snippets, thanks!

File: sdk/keyvault/azure-keyvault-certificates/azure/keyvault/certificates/client.py (hunk 1 above)
Comment (iscai-msft) [reply sdk/keyvault/azure-keyvault-certificates/azure/keyvault/certificates/client.py]: nice, got rid of these checks in the clients

File: sdk/keyvault/azure-keyvault-certificates/azure/keyvault/certificates/models.py (hunk 5 above)
Comment (iscai-msft) [reply sdk/keyvault/azure-keyvault-certificates/azure/keyvault/certificates/models.py]: I'm going to opening up a different pr with increased test coverage for certificate policy

File: sdk/keyvault/azure-keyvault-certificates/azure/keyvault/certificates/enums.py (hunk 2 above)
Comment (iscai-msft) [reply sdk/keyvault/azure-keyvault-certificates/azure/keyvault/certificates/enums.py]: no it doesn't, i forgot to check. I've tried to condense it into one line

File: sdk/keyvault/azure-keyvault-certificates/azure/keyvault/certificates/enums.py [hunk 6]
```diff
@@ -50,3 +50,10 @@ class KeyCurveName(str, Enum):
     p_384 = "P-384"  #: The NIST P-384 elliptic curve, AKA SECG curve SECP384R1.
     p_521 = "P-521"  #: The NIST P-521 elliptic curve, AKA SECG curve SECP521R1.
     p_256_k = "P-256K"  #: The SECG SECP256K1 elliptic curve.
+
+
+class WellKnownIssuerNames(str, Enum):
+    """Collection of well-known issuer names"""
+
+    Self = "Self"  #: Use this issuer for a self-signed certificate
+    Unknown = "Unknown"  #: Here you must get the cert yourself from your issuer and call :func:`merge_certificate`.
```
Comment (chlowell) [inline sdk/keyvault/azure-keyvault-certificates/azure/keyvault/certificates/enums.py]: Does sphinx resolve this link?

**Pull request merged**